core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -0,0 +1,251 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool.docker
+
+import scala.concurrent.Future
+import spray.json.JsObject
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext
+import whisk.core.container.HttpUtils
+import whisk.core.entity.size._
+import scala.util.Success
+import scala.util.Failure
+import java.time.Instant
+import whisk.core.container.RunResult
+import whisk.core.container.Interval
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+import whisk.common.TransactionId
+import whisk.common.Logging
+import whisk.common.LoggingMarkers
+import whisk.core.entity.ByteSize
+import java.nio.charset.StandardCharsets
+import whisk.http.Messages
+import whisk.core.entity.ActivationResponse
+import whisk.core.containerpool.BlackboxStartupError
+import whisk.core.containerpool.WhiskContainerStartupError
+import whisk.core.containerpool.InitializationError
+import whisk.core.containerpool.Container
+import whisk.core.invoker.ActionLogDriver
+
+object DockerContainer {
+    /**
+     * Creates a container running on a docker daemon.
+     *
+     * @param transid transaction creating the container
+     * @param image image to create the container from
+     * @param userProvidedImage whether the image is provided by the user
+     *     or is an OpenWhisk provided image
+     * @param memory memorylimit of the container
+     * @param cpuShares sharefactor for the container
+     * @param environment environment variables to set on the container
+     * @param network network to launch the container in
+     * @param name optional name for the container
+     * @return a container
+     */
+    def create(transid: TransactionId,
+               image: String,
+               userProvidedImage: Boolean = false,
+               memory: ByteSize = 256.MB,
+               cpuShares: Int = 0,
+               environment: Map[String, String] = Map(),
+               network: String = "bridge",
+               name: Option[String] = None)(
+                   implicit docker: DockerApiWithFileAccess, runc: RuncApi, ec: ExecutionContext, log: Logging) = {
+        implicit val tid = transid
+
+        val environmentArgs = (environment + ("SERVICE_IGNORE" -> true.toString)).map {
+            case (key, value) => Seq("-e", s"$key=$value")
+        }.flatten
+
+        val args = Seq(
+            "--cap-drop", "NET_RAW",
+            "--cap-drop", "NET_ADMIN",
+            "--ulimit", "nofile=1024:1024",
+            "--pids-limit", "1024", // OW PR 2119
+            "--cpu-shares", cpuShares.toString,
+            "--memory", s"${memory.toMB}m",
+            "--memory-swap", s"${memory.toMB}m",
+            "--network", network) ++
+            environmentArgs ++
+            name.map(n => Seq("--name", n)).getOrElse(Seq.empty)
+
+        val pulled = if (userProvidedImage) {
+            docker.pull(image).recoverWith {
```

Comment:
I agree - `docker pull` takes potentially very long and may block a single Docker container for a long time. It's reasonable to let the operation fail if pulling an image takes too long.

- What would be a reasonable timeout duration for `docker pull`? Should we re-use the same timeout we apply to initialize / run? Or do we need a general `docker pull` timeout?
- At the moment, I'm building a prototype with a timeout mechanism based on `akka.pattern.after`. Unfortunately, said `after()` needs a scheduler - most likely implying that we need the proper actor system at this point. So instead of passing the `ExecutionContext`, we need the whole `ActorSystem` and use `ActorSystem.scheduler` and `ActorSystem.dispatcher`.
- Said approach with `akka.pattern.after` used in `DockerContainer.create()` will let the creation fail with a timeout if it takes too long - but won't actually abort the running `docker pull` command process. Our current command execution approach does not support aborting a running process. We may think about providing a process executor that observes timeouts and abort execution upon timeout.